tools/clang/lib/SPIRV/SpirvBuilder.cpp

```diff
@@ -86,5 +87,576 @@ void SpirvBuilder::setContinueTarget(SpirvBasicBlock *continueLabel) {
+  if (isSparse) {
+    requireCapability(spv::Capability::SparseResidency);
+  }
+
+  spv::Op op =
+      doImageFetch
+          ? (isSparse ? spv::Op::OpImageSparseFetch : spv::Op::OpImageFetch)
+          : (isSparse ? spv::Op::OpImageSparseRead : spv::Op::OpImageRead);
+
+  if (!doImageFetch) {
+    requireCapability(
+        TypeTranslator::getCapabilityForStorageImageReadWrite(imageType));
+  }
+
+  auto *fetchOrReadInst = new (context) SpirvImageOp(
+      op, texelType, /*id*/ 0, loc, image, coordinate, mask,
+      /*dref*/ nullptr, /*bias*/ nullptr, lod, /*gradDx*/ nullptr,
+      /*gradDy*/ nullptr, constOffset, varOffset, constOffsets, sample);
+  insertPoint->addInstruction(fetchOrReadInst);
+
+  if (isSparse) {
+    // Write the Residency Code
+    const auto status =
+        createCompositeExtract(astContext.UnsignedIntTy, fetchOrReadInst, {0});
+    createStore(residencyCode, status, loc);
+    // Extract the real result from the struct
+    return createCompositeExtract(texelType, fetchOrReadInst, {1});
+  }
+
+  return fetchOrReadInst;
+}
+
+void SpirvBuilder::createImageWrite(QualType imageType, SpirvInstruction *image,
+                                    SpirvInstruction *coord,
+                                    SpirvInstruction *texel,
+                                    SourceLocation loc) {
+  assert(insertPoint && "null insert point");
+  requireCapability(
+      TypeTranslator::getCapabilityForStorageImageReadWrite(imageType));
+  auto *writeInst = new (context) SpirvImageOp(
+      spv::Op::OpImageWrite, imageType, /*id*/ 0, loc, image, coord,
+      spv::ImageOperandsMask::MaskNone,
+      /*dref*/ nullptr, /*bias*/ nullptr, /*lod*/ nullptr, /*gradDx*/ nullptr,
+      /*gradDy*/ nullptr, /*constOffset*/ nullptr, /*varOffset*/ nullptr,
+      /*constOffsets*/ nullptr, /*sample*/ nullptr, /*minLod*/ nullptr,
+      /*component*/ nullptr, texel);
+  insertPoint->addInstruction(writeInst);
+}
+
+SpirvInstruction *SpirvBuilder::createImageGather(
+    QualType texelType, QualType imageType, SpirvInstruction *image,
+    SpirvInstruction *sampler, bool isNonUniform, SpirvInstruction *coordinate,
+    SpirvInstruction *component, SpirvInstruction *compareVal,
+    SpirvInstruction *constOffset, SpirvInstruction *varOffset,
+    SpirvInstruction *constOffsets, SpirvInstruction *sample,
+    SpirvInstruction *residencyCode, SourceLocation loc) {
+  assert(insertPoint && "null insert point");
+
+  if (residencyCode) {
+    requireCapability(spv::Capability::SparseResidency);
+  }
+
+  // An OpSampledImage is required to do the image sampling.
+  auto *sampledImage =
+      new (context) SpirvSampledImage(imageType, /*id*/ 0, loc, image, sampler);
+  insertPoint->addInstruction(sampledImage);
+
+  if (isNonUniform) {
+    // The sampled image will be used to access resource's memory, so we need
+    // to decorate it with NonUniformEXT.
+    // TODO: decorateNonUniformEXT(sampledImgId);
```

Comment:
This will be re-enabled in my next commit. (when decorate* methods are implemented)